File: classical_methods.cc
#include <iostream>
#include <cmath>
#include <unordered_map>
#include <string>

// Function to calculate the bearing capacity factors using Terzaghi's method
std::unordered_map<std::string, double> terzaghi(double phi) {
    std::unordered_map<std::string, double> factors;

    double exponent = exp(M_PI * (0.75 - (phi / 360.0)) * tan(phi * M_PI / 180.0));
    double numerator_q = exponent * exponent;
    double denominator_q = 2 * pow(cos((45.0 + (phi / 2.0)) * M_PI / 180.0), 2);
    double Nq = numerator_q / denominator_q;

    double Nc = (phi > 0) ? ((Nq - 1) * (1 / tan(phi * M_PI / 180.0))) : 5.71;

    double numerator_g = 2 * (Nq + 1) * tan(phi * M_PI / 180.0);
    double denominator_g = 1 + (0.4 * sin(4 * phi * M_PI / 180.0));
    double Ngamma = numerator_g / denominator_g;

    factors["Nc"] = Nc;
    factors["Nq"] = Nq;
    factors["Ngamma"] = Ngamma;

    return factors;
}

// Function to calculate the bearing capacity factors using Meyerhof's method
std::unordered_map<std::string, double> meyerhof(double phi) {
    std::unordered_map<std::string, double> factors;

    double Nq = exp(M_PI * tan(phi * M_PI / 180.0)) * pow(tan((45.0 + (phi / 2.0)) * M_PI / 180.0), 2);

    double Nc = (phi > 0) ? ((Nq - 1) * (1 / tan(phi * M_PI / 180.0))) : 5.14;

    double Ngamma = (Nq - 1) * tan(phi * 1.4 * M_PI / 180.0);

    factors["Nc"] = Nc;
    factors["Nq"] = Nq;
    factors["Ngamma"] = Ngamma;

    return factors;
}

// Function to calculate the bearing capacity factors using Vesic's method
std::unordered_map<std::string, double> vesic(double phi) {
    std::unordered_map<std::string, double> factors;

    double Nq = exp(M_PI * tan(phi * M_PI / 180.0)) * pow(tan((45.0 + (phi / 2.0)) * M_PI / 180.0), 2);

    double Nc = (phi > 0) ? ((Nq - 1) * (1 / tan(phi * M_PI / 180.0))) : 5.14;

    double Ngamma = 2 * (Nq + 1) * tan(phi * M_PI / 180.0);

    factors["Nc"] = Nc;
    factors["Nq"] = Nq;
    factors["Ngamma"] = Ngamma;

    return factors;
}

// Function to calculate the bearing capacity factors using Hansen's method
std::unordered_map<std::string, double> hansen(double phi) {
    std::unordered_map<std::string, double> factors;

    double Nq = exp(M_PI * tan(phi * M_PI / 180.0)) * pow(tan((45.0 + (phi / 2.0)) * M_PI / 180.0), 2);

    double Nc = (phi > 0) ? ((Nq - 1) * (1 / tan(phi * M_PI / 180.0))) : 5.14;

    double Ngamma = 1.5 * (Nq - 1) * tan(phi * M_PI / 180.0);

    factors["Nc"] = Nc;
    factors["Nq"] = Nq;
    factors["Ngamma"] = Ngamma;

    return factors;
}

// Function to calculate the bearing capacity factors following Eurocode 7 (EC7)
std::unordered_map<std::string, double> EC7(double phi) {
    std::unordered_map<std::string, double> factors;

    double Nq = exp(M_PI * tan(phi * M_PI / 180.0)) * pow(tan((45.0 + (phi / 2.0)) * M_PI / 180.0), 2);

    double Nc = (phi > 0) ? ((Nq - 1) * (1 / tan(phi * M_PI / 180.0))) : 5.14;

    double Ngamma = 2 * (Nq - 1) * tan(phi * M_PI / 180.0);

    factors["Nc"] = Nc;
    factors["Nq"] = Nq;
    factors["Ngamma"] = Ngamma;

    return factors;
}

File: classical_methods.rs
use std::f64::consts::PI;
use std::collections::HashMap;

// Function to calculate the bearing capacity factors using Terzaghi's method
fn terzaghi(phi: f64) -> HashMap<String, f64> {
    let mut factors = HashMap::new();

    let exponent = (PI * (0.75 - (phi / 360.0)) * (phi * PI / 180.0).tan()).exp();
    let numerator_q = exponent.powi(2);
    let denominator_q = 2.0 * ((45.0 + (phi / 2.0)) * PI / 180.0).cos().powi(2);
    let nq = numerator_q / denominator_q;

    let nc = if phi > 0.0 {
        (nq - 1.0) * (1.0 / (phi * PI / 180.0).tan())
    } else {
        5.71
    };

    let numerator_g = 2.0 * (nq + 1.0) * (phi * PI / 180.0).tan();
    let denominator_g = 1.0 + (0.4 * (4.0 * phi * PI / 180.0).sin());
    let ngamma = numerator_g / denominator_g;

    factors.insert("Nc".to_string(), nc);
    factors.insert("Nq".to_string(), nq);
    factors.insert("Ngamma".to_string(), ngamma);

    factors
}

// Function to calculate the bearing capacity factors using Meyerhof's method
fn meyerhof(phi: f64) -> HashMap<String, f64> {
    let mut factors = HashMap::new();

    let nq = (PI * (phi * PI / 180.0).tan()).exp() * ((45.0 + (phi / 2.0)) * PI / 180.0).tan().powi(2);

    let nc = if phi > 0.0 {
        (nq - 1.0) * (1.0 / (phi * PI / 180.0).tan())
    } else {
        5.14
    };

    let ngamma = (nq - 1.0) * (phi * 1.4 * PI / 180.0).tan();

    factors.insert("Nc".to_string(), nc);
    factors.insert("Nq".to_string(), nq);
    factors.insert("Ngamma".to_string(), ngamma);

    factors
}

// Function to calculate the bearing capacity factors using Vesic's method
fn vesic(phi: f64) -> HashMap<String, f64> {
    let mut factors = HashMap::new();

    let nq = (PI * (phi * PI / 180.0).tan()).exp() * ((45.0 + (phi / 2.0)) * PI / 180.0).tan().powi(2);

    let nc = if phi > 0.0 {
        (nq - 1.0) * (1.0 / (phi * PI / 180.0).tan())
    } else {
        5.14
    };

    let ngamma = 2.0 * (nq + 1.0) * (phi * PI / 180.0).tan();

    factors.insert("Nc".to_string(), nc);
    factors.insert("Nq".to_string(), nq);
    factors.insert("Ngamma".to_string(), ngamma);

    factors
}

// Function to calculate the bearing capacity factors using Hansen's method
fn hansen(phi: f64) -> HashMap<String, f64> {
    let mut factors = HashMap::new();

    let nq = (PI * (phi * PI / 180.0).tan()).exp() * ((45.0 + (phi / 2.0)) * PI / 180.0).tan().powi(2);

    let nc = if phi > 0.0 {
        (nq - 1.0) * (1.0 / (phi * PI / 180.0).tan())
    } else {
        5.14
    };

    let ngamma = 1.5 * (nq - 1.0) * (phi * PI / 180.0).tan();

    factors.insert("Nc".to_string(), nc);
    factors.insert("Nq".to_string(), nq);
    factors.insert("Ngamma".to_string(), ngamma);

    factors
}

// Function to calculate the bearing capacity factors following Eurocode 7 (EC7)
fn ec7(phi: f64) -> HashMap<String, f64> {
    let mut factors = HashMap::new();

    let nq = (PI * (phi * PI / 180.0).tan()).exp() * ((45.0 + (phi / 2.0)) * PI / 180.0).tan().powi(2);

    let nc = if phi > 0.0 {
        (nq - 1.0) * (1.0 / (phi * PI / 180.0).tan())
    } else {
        5.14
    };

    let ngamma = 2.0 * (nq - 1.0) * (phi * PI / 180.0).tan();

    factors.insert("Nc".to_string(), nc);
    factors.insert("Nq".to_string(), nq);
    factors.insert("Ngamma".to_string(), ngamma);

    factors
}
